### ISSUE TITLE
[TimePicker] `onError` not trigerring

### ISSUE THREAD

**Issue by kkaplita** [mui/mui-x#6402]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Hi!

I have some little problem. `TimePicker`'s `onError` prop not firing when entered value is wrong. I can enter `99:99` for example and validation function not triggering in any case.

### Expected behavior 🤔

When enter wrong value, `onError` function should trigger error.

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by yaredtsy** [comment]
@kkaplita check this example https://codesandbox.io/s/practical-ardinghelli-0jtct9?file=/demo.tsx . 
you can also check if the dates are valid using.

`<TimePicker
       ......
        onChange={(newValue) => {
          console.log(newValue?.isValid());
        }}
 />`

**Comment by flaviendelangle** [comment]
The validation  logic is to trigger `onError` each time the error changes.

For instance  if you start with an empty `TimePicker` and type `99:99` (by pressing 4 times `9`, you should have an `invalidDate` error after the first stroke, but after that the error is always `invalidDate` so we don't fire it again.

For the current components, I don't see how to drastically improve this behavior without calling `onError` everytime something changes even if the error remains the same. But that would cause a massive flood of useless calls, which we clearly don't want.

For the new components ([TimeField](https://next.mui.com/x/react-date-pickers/time-field/) for instance), you won't be able to type `99` in the minute or hour sections.
And just like today `onError` will be called when pressing the first `9` because the date is not valid yet.
That makes me think that some people will probably want a mode where the error is only fired when all sections are filled.


___

Long story short, that is pretty much the expected behavior.
If the scenario I am describing does not match yours (the 4 `9` key strokes), please describe in detail the yours so I can reproduce it and see if there is a bug.
